cunumeric/deferred.py

```diff
@@ -305,37 +305,235 @@ def get_scalar_array(self):
+        # the transformation, we need to return a copy
+        tuple_of_arrays = ()
+
+        # First, we need to check if transpose is needed
+        for dim, k in enumerate(key):
+            if np.isscalar(k) or isinstance(k, NumPyThunk):
+                if start_index == -1:
+                    start_index = dim
+                transpose_indices += (dim,)
+                transpose_needed = transpose_needed or ((dim - last_index) > 1)
+                last_index = dim
+
+        if transpose_needed:
+            start_index = 0
+            post_indices = tuple(
+                i for i in range(store.ndim) if i not in transpose_indices
+            )
+            key_transpose_indices = transpose_indices
+            transpose_indices += post_indices
+            post_indices = tuple(
+                i for i in range(len(key)) if i not in key_transpose_indices
+            )
+            key_transpose_indices += post_indices
+            store = store.transpose(transpose_indices)
+
+            key = tuple(key[i] for i in key_transpose_indices)
+
+        for d, k in enumerate(key):
+            dim = d
+            if np.isscalar(k):
+                if k < 0:
+                    k += store.shape[dim + shift]
+                store = store.project(dim + shift, k)
+                shift -= 1
+            elif k is np.newaxis:
+                store = store.promote(dim + shift, 1)
+            elif isinstance(k, slice):
+                store = store.slice(dim + shift, k)
+            elif isinstance(k, NumPyThunk):
+                if not isinstance(key, DeferredArray):
+                    k = self.runtime.to_deferred_array(k)
+                if k.dtype == np.bool:
+                    if k.shape[0] != store.shape[dim + shift]:
+                        raise ValueError(
+                            "shape of boolean index did not match "
+                            "indexed array "
+                        )
+                    # in case of the mixed indises we all nonzero
+                    # for the bool array
+                    k = k.nonzero()
+                    tuple_of_arrays += k
+                else:
+                    tuple_of_arrays += (k,)
+            else:
+                raise TypeError(
+                    "Unsupported entry type passed to advanced ",
+                    "indexing operation",
+                )
+        if store.transformed:
+            # after store is transformed we need to to return a copy of
+            # the store since Copy operation can't be done on
+            # the store with transformation
+            rhs, store = self._copy_store(store)
+
+        if len(tuple_of_arrays) <= rhs.ndim and rhs.ndim > 1:
+            output_arr = rhs._zip_indices(start_index, tuple_of_arrays)
+            return True, rhs, output_arr
+        elif len(tuple_of_arrays) == 1 and rhs.ndim == 1:
+            return True, rhs, tuple_of_arrays[0]
```

Comment:
I think you need to check if `tuple_of_arrays[0]` is transformed at this point, and `_copy_store` if so.

Reply:
thank you